src/lib/reasoners/bitv_rel.ml

```diff
@@ -1877,94 +1951,144 @@ let simplify_all uf eqs touched (dom, idom) =
         ) c.value (dom, idom)
     ) to_add (dom, idom)
 
+(* Holds the queues for propagation.
+
+   Currently we use two separate queues for single-domain propagators (i.e.
+   propagators that only use one domain) and for multi-domain propagators
+   (i.e. propagators that propagate information across different domains,
+   e.g. between bitlists and intervals).
+
+   TODO(bclement): generalize to an arbitrary queue combination (and possibly
+   infer the queue configuration from the schedule?). *)
+type queues =
+  { to_simplify : Ex.t HC.t
+  ; propagation_queue : Any_propagator.Queue.t
+  ; consistency_queue : Any_propagator.Queue.t }
+
+let state uf idom bdom =
+  let to_simplify = HC.create 17 in
+  let propagation_queue = Any_propagator.Queue.create 17 in
+  let consistency_queue = Any_propagator.Queue.create 17 in
+  let events
+      constraint_propagator
+      structural_propagator
+      consistency_propagator
+    =
+    let evt_atomic_change r =
+      Any_propagator.Queue.push propagation_queue structural_propagator r;
+      Any_propagator.Queue.push consistency_queue consistency_propagator r
+    and evt_composite_change r =
+      Any_propagator.Queue.push propagation_queue structural_propagator r;
+      Any_propagator.Queue.push consistency_queue consistency_propagator r
+    and evt_watch_trigger c =
+      HC.replace to_simplify c.value c.explanation;
+      Any_propagator.Queue.push propagation_queue constraint_propagator c
+    in
+    { Domains_intf.evt_atomic_change
+    ; evt_composite_change
+    ; evt_watch_trigger
+    }
+  in
+
+  let bitlist_events =
+    events
+      bitlist_constraint_propagator
+      bitlist_structural_propagator
+      interval_from_bitlist_propagator
+  and interval_events =
+    events
+      interval_constraint_propagator
+      interval_structural_propagator
+      bitlist_from_interval_propagator
+  in
+  let bvars = Bitlist_domains.parents bdom in
+  let ivars = Interval_domains.parents idom in
+
+  let bdom = Bitlist_domains.edit ~events:bitlist_events bdom in
+  let idom = Interval_domains.edit ~events:interval_events idom in
+
+  let uf_bdom = Bitlist_domains.Ephemeral.canon uf bdom in
+  let uf_idom = Interval_domains.Ephemeral.canon uf idom in
+  { to_simplify ; propagation_queue ; consistency_queue },
+  { intervals = idom
+  ; intervals_uf = uf_idom
+  ; interval_variables = ivars
+  ; interval_changed = HX.create 17
+  ; bitlists = bdom
+  ; bitlists_uf = uf_bdom
+  ; bitlist_variables = bvars
+  ; bitlist_changed = HX.create 17 }
+
+(* Schedule specifications *)
+type schedule =
+  | Single of Any_propagator.Queue.t
+  (** Propagate a single constraint from the given queue. *)
+  | Round_robin of (schedule * int) array
+  (** Performs [n] propagations from the provided schedules in round-robin
+      fashion the coefficient adjusts how many propagations are performed
+      before moving on to the next schedule. *)
```

Comment:
Not yet. I think I initially wanted to experiment with it then didn't. Also using "round-robin" here is somewhat incorrect since we don't cycle. I renamed and simplified the constructors.